### PR TITLE
Base RaspaBaseWorkChain on BaseRestartWorkChain from aiida-core.

### PR DESCRIPTION
**Pull request by yakutovicha**

### CODE REVIEW

**Comment by yakutovicha** [comment]
@pzarabadip I noticed a few things here:

* example_base_workchain_gcmc_2comp.py did not converge at the end.

* examples/workchains/example_base_workchain_widom_1comp.py did not finish after 5 iterations.

* examples/workchains/example_base_workchain_gemc_1comp.py does not call handlers.

Can you have a look, please? Or should we just merge this PR and make a separate issue for those problems?

**Comment by yakutovicha** [comment]
this can only be merged after https://github.com/aiidateam/aiida-core/pull/3786.

**Comment by pzarabadip** [comment]
Sure @yakutovicha. I will check them and will let you know.

**Comment by pzarabadip** [comment]
> @pzarabadip I noticed a few things here:
> 
> * example_base_workchain_gcmc_2comp.py did not converge at the end.
> * examples/workchains/example_base_workchain_widom_1comp.py did not finish after 5 iterations.
> * examples/workchains/example_base_workchain_gemc_1comp.py does not call handlers.
> 
> Can you have a look, please? Or should we just merge this PR and make a separate issue for those problems?

I just checked these examples @yakutovicha 
The reason that these are not being converged is that `additional_init_cycle` and `additional_prod_cycle` are hard-coded to `0` in handlers. 
It would be great if we can make them as well as `conv_threshold` user-defined. So, user can pass them to builder for instance like:
```
builder.handler_overrides = Dict(dict={
        'check_gcmc_convergence': True,
        'conv_threshold': 0.2,
        'additional_init_cycle': 1000,
        'additional_prod_cycle': 2000
    })
```
I will mark the parts that needs to change separately.

File: aiida_raspa/workchains/base.py [hunk 1]
```diff
@@ -45,18 +45,162 @@ def report_error_handled(self, calculation, action):
...
         self.ctx.inputs.parent_folder = calculation.outputs.remote_folder
+        return ProcessHandlerReport(False)
+
+    @process_handler(priority=400, enabled=False)
+    def check_widom_convergence(self, calculation):
+        """Checks whether a Widom particle insertion is converged. The check is based on the
+        error bar of the Henry coefficient."""
+
+        conv_threshold = 0.1
+        additional_cycle = 0
```
Comment (pzarabadip) [inline aiida_raspa/workchains/base.py]: Here, it needs to change to accept `conv_threshold` and `additional_cycle` as arguments.

File: aiida_raspa/workchains/base.py [hunk 2]
```diff
@@ -45,18 +45,162 @@ def report_error_handled(self, calculation, action):
...
+            return ProcessHandlerReport(False)
+
+        return None
+
+    @process_handler(priority=410, enabled=False)
+    def check_gcmc_convergence(self, calc):
+        """Checks whether a GCMC calc is converged. Checking is based on the error bar on average loading."""
+        conv_threshold = 0.1
+        additional_init_cycle = 0
+        additional_prod_cycle = 0
```
Comment (pzarabadip) [inline aiida_raspa/workchains/base.py]: Here, it needs to change to accept `conv_threshold` and `additional_init_cycle` and `additiona_prod_cycle` as arguments.

File: aiida_raspa/workchains/base.py [hunk 3]
```diff
@@ -45,18 +45,162 @@ def report_error_handled(self, calculation, action):
...
+        return None
+
+    @process_handler(priority=410, enabled=False)
+    def check_gemc_convergence(self, calc):
+        """Checks whether a GCMC calc is converged. Checking is based on the error bar on average loading which is
+        average number of molecules in each simulation box."""
+
+        conv_threshold = 0.1
+        additional_init_cycle = 0
+        additional_prod_cycle = 0
```
Comment (pzarabadip) [inline aiida_raspa/workchains/base.py]: Here, it needs to change to accept `conv_threshold` and `additional_init_cycle` and `additiona_prod_cycle` as arguments.

File: examples/workchains/example_base_workchain_gcmc_2comp.py [hunk 4]
```diff
@@ -89,10 +89,8 @@ def example_base_workchain_gcmc(raspa_code):
         "irmof_1_xenon": block_pocket_node2,
     }
 
-    # Add fixers that could handle physics-related problems.
-    builder.fixers = {
-        'fixer_001': ('aiida_raspa.utils', 'check_gcmc_convergence', 0.10, 2000, 2000),
-    }
+    builder.handler_overrides = Dict(dict={'check_gcmc_convergence': True
+                                          })  # Enable gcmc convergence handler disabled by default.
```
Comment (pzarabadip) [inline examples/workchains/example_base_workchain_gcmc_2comp.py]: After fixing the above-mentioned points, we need to pass the arguments here and in the other examples.

File: aiida_raspa/workchains/base.py (hunk 1 above)
Comment (yakutovicha) [reply aiida_raspa/workchains/base.py]: We cannot, unfortunately. Handlers do not accept input parameters.

**Comment by yakutovicha** [comment]
>  It would be great if we can make them as well as `conv_threshold` user-defined. So, user can pass them to builder for instance like:

Unfortunately, handlers do not accept additional input parameters..

File: aiida_raspa/workchains/base.py (hunk 1 above)
Comment (pzarabadip) [reply aiida_raspa/workchains/base.py]: Oh, I see. Then, I think the solution would be keeping the `conv_threshold` as it is and for the `additiona_cycle` I suggest to have it either 
* as the product of a constant value (let's say 0.5) by initial value set by the user
* or a fixed value.
I, myself, vote for the first option. 
We can later do a benchmark to investigate more and modify the the code that would reflect handling of majority of cases.

**Pull request merged**